proxy/proxy/httpserver.py

```diff
@@ -0,0 +1,171 @@
+"""proxy server"""
+from http import HTTPStatus
+from http.server import ThreadingHTTPServer, BaseHTTPRequestHandler
+import json
+import logging
+import os
+import ssl
+import zlib
+import requests
+
+logging.basicConfig(level=logging.DEBUG)
+
+HOST = "127.0.0.1"
+PORT = 8443
+TIMEOUT = 30
+
+TEST_DST_PROTOCOL = "http"
+TEST_GATEWAY_URL = "127.0.0.1:60000"
+dst_protocol = os.environ.get("DST_PROTOCOL", TEST_DST_PROTOCOL)
+gateway_url = os.environ.get("GATEWAY_URL", TEST_GATEWAY_URL)
+
+
+class ProxyRequestHandler(BaseHTTPRequestHandler):
+    """Proxy Request Handler."""
+
+    def gzip_encode(self, content):
+        """gzip encode"""
+        gzip_compress = zlib.compressobj(9, zlib.DEFLATED, zlib.MAX_WBITS | 16)
+        data = gzip_compress.compress(content) + gzip_compress.flush()
+        return data
+
+    def handle_response(self, resp):
+        """handle response."""
+        self.send_response(resp.status_code)
+        if resp.content and len(resp.content) != 0:
+            content = resp.content
+            logging.debug("Recieved response content: %s", content.decode("utf-8"))
+            if (
+                "content-encoding" in resp.headers
+                and resp.headers["content-encoding"] == "gzip"
+            ):
+                content = self.gzip_encode(content)
+            if (
+                "Transfer-Encoding" in resp.headers
+                and resp.headers["Transfer-Encoding"] == "chunked"
+            ):
+                content = (
+                    bytes(f"{len(content):x}", "utf-8")
+                    + b"\x0d\x0a"
+                    + content
+                    + b"\x0d\x0a0\x0d\x0a\x0d\x0a"
+                )
+        for header in resp.headers:
+            self.send_header(header, resp.headers[header])
+            logging.debug("header: %s, %s", header, resp.headers[header])
+        self.end_headers()
+        if resp.content and len(resp.content) != 0:
+            self.wfile.write(content)
+            logging.debug("Sending response content: %s", content)
+
+    def handle_one_request(self):
+        """Handle a single HTTP request."""
+
+        try:
+            logging.debug("New request")
+            self.raw_requestline = (  # pylint: disable=attribute-defined-outside-init
+                self.rfile.readline(65537)
```

Comment:
What are we doing here, @akihikokuroda ?

Reply:
It's reading the request contents.

Reply:
for what is this number? Can this number change?

Reply:
I don't remember where I picked this value.  It's just a big enough number for the request data size.

Reply:
This data size can change depending of the input of the user? Because something that we are discussing is that for the uses cases that we are going to need to handle we can expect a couple of hundred of megabytes in some user inputs.

Reply:
I looked this line again.  It is reading the request uri and headers.  It doesn't grow  by the user input.  The user data in the POST requests are read in the `do_POST` with the `Content-Length` header value.

Reply:
Thank you so much, Aki. It would be nice if we could clarify from where this number comes from 🙏

Reply:
I found the RFC that has the recommended value for this.  I made a constant for it and put the reference in the comment.

Reply:
Thank you @akihikokuroda